server/shared/src/main/scala/org/http4s/server/middleware/authentication/NonceKeeperF.scala

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright 2014 http4s.org
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.http4s.server.middleware.authentication
+
+import cats.effect.Async
+import cats.effect.Ref
+import cats.effect.std.Random
+import cats.effect.std.Semaphore
+import cats.syntax.all._
+
+import java.util.LinkedHashMap
+import java.{util => ju}
+import scala.concurrent.duration.Duration
+import scala.concurrent.duration.FiniteDuration
+
+private[authentication] object NonceKeeperF {
+  def apply[F[_]](
+      staleTimeout: Duration,
+      nonceCleanupInterval: Duration,
+      bits: Int,
+  )(implicit F: Async[F]): F[NonceKeeperF[F]] = for {
+    // This semaphore controls who has access to `nonces` during stale nonce eviction. This must never be set above one.
+    semaphore <- Semaphore[F](1)
+    current <- F.monotonic
+    lastCleanupMillis <- Ref[F].of(current)
+    nonces = new LinkedHashMap[String, NonceF[F]]
+    random <- Random.javaSecuritySecureRandom[F]
```

Comment:
I see in CE3 this is implemented as:
```scala
  def javaSecuritySecureRandom[F[_]: Sync]: F[Random[F]] =
    Sync[F].delay(new java.security.SecureRandom).flatMap(r => javaUtilRandom(r))
```

We can do the same thing and unblock JS:
```scala
  def javaSecuritySecureRandom[F[_]: Sync]: F[Random[F]] =
    Sync[F].delay(new org.http4s.crypto.unsafe.SecureRandom).flatMap(r => javaUtilRandom(r))
```

Reply:
Oops.  I didn't remember making [6753dd0](https://github.com/http4s/http4s/pull/6178/commits/6753dd0265c41a20b6f4e01dedef1a1580974508) part of this.  But it's a tight spot either way:

* If we do it, we want a consensus on `cats.effect.std.SecureRandom` before we release.
* If we don't do it, we can't easily introduce `Random` to the API later without another crap overload.

Reply:
I'm going to prune that commit because there's a discussion of it on another draft.